### ISSUE TITLE
Better IDA Pro integration

### ISSUE THREAD

**Issue by disconnect3d** [pwndbg/pwndbg#132]
Currently it is not possible to integrate Ida Pro rpc server hosted on Windows host machine with pwndbg on a Linux vm guest without some hacky changes (see below).

The problem is there because of:
* pwndbg by default connects to localhost: https://github.com/pwndbg/pwndbg/blob/master/pwndbg/ida.py#L49
* the ida server script tries to save database into a path that is invalid on Windows - https://github.com/pwndbg/pwndbg/blob/master/ida_script.py#L19 - here is a proof: https://gyazo.com/651e3fda29ecac06fb5734554c537961
* this datetime path is also problematic in here - https://github.com/pwndbg/pwndbg/blob/master/pwndbg/commands/ida.py#L94 - we probably need a possibility to provide path to the database on both machines OR send the database file between them?

I have managed to make such integration working by making such changes:
```
diff --git a/pwndbg/commands/ida.py b/pwndbg/commands/ida.py
index b2c095d..cd6f69c 100644
--- a/pwndbg/commands/ida.py
+++ b/pwndbg/commands/ida.py
@@ -88,15 +88,20 @@ def save_ida():
         os.mkdir(backups)
 
     basename, ext = os.path.splitext(basename)
-    basename += '-%s' % datetime.datetime.now().isoformat()
-    basename += ext
+    #basename += '-%s' % datetime.datetime.now().isoformat()
+    #basename += ext + '.pwndbg'
+    print("BASENAME ='%s'" % basename)
 
     # Windows doesn't like colons in paths
-    basename = basename.replace(':','_')
+    #basename = basename.replace(':','_')
 
-    full_path = os.path.join(backups, basename)
+    #full_path = os.path.join(backups, basename)
+    full_path = basename  + ext
 
     pwndbg.ida.SaveBase(full_path)
+    full_path = full_path.replace("D:", "/d")
+    full_path = full_path.replace("\\", "/")
+    print("FULLPATH = '%s'" % full_path)
 
     data = open(full_path, 'rb').read()
 
diff --git a/pwndbg/ida.py b/pwndbg/ida.py
index 82311f7..528cb74 100644
--- a/pwndbg/ida.py
+++ b/pwndbg/ida.py
@@ -43,7 +43,8 @@ xmlrpclib.Marshaller.dispatch[type(0)] = lambda _, v, w: w("<value><i8>%d</i8></
 
 def setPort(port):
     global _ida
-    _ida = xmlrpclib.ServerProxy('http://localhost:%s' % port)
+    _ida = xmlrpclib.ServerProxy('http://192.168.56.1:%s' %  port)
     try:
         _ida.here()
     except socket.error as e:
```
(My "D:/" drive is accessible on linux by "/d/").
This also required to change the ip the ida pro rpc server binds to.

However we probably need a way to configure:
* IDA Pro server ip address
* Path mappings between server (machine with ida pro server) and client (machine with pwndbg)

Is there any place for such configurations? I can work on that.

The next thing to consider are features, e.g. it seems to me it is not possible to set a breakpoint on function name that is synchronized from Ida database (see https://gyazo.com/b77cd34847fa2986b15d6d5953840875). But that should probably be another issue.

**Comment by zachriggle** [comment]
It is *literally* impossible to introduce new native symbols into IDA.  All IDA symbols appear as "native" within Pwndbg functionality, of which breakpoints (and e.g. "X/") are not.

There is a helper function which works around this.

$ break *$ida("symbol")

**Comment by zachriggle** [comment]
The bit about path mappings and IPs I don't want to have to deal with.  Just use IDA under Wine, or socat or SSH for port forwarding.

Wine (or native IDA) is your friend ;)

**Comment by disconnect3d** [comment]
> It is literally impossible to introduce new native symbols into IDA.

Did you mean gdb?

Is there a possibility to override a gdb command? If so that could have been a solution (a hackish one... but probably worth it?).

> The bit about path mappings and IPs I don't want to have to deal with. Just use IDA under Wine, or socat or SSH for port forwarding.

I guess only the IDA under Wine would work because of the paths that are different between machines...

Would you accept a MR that would add a configuration file for such integration and use it?

**Comment by zachriggle** [comment]
Sure, I'll add any configs you'd like, but I don't have a Windows VM to
suds everything out and make sure things work.

If you'd like to make a pull request which causes the current behavior to
be the default, and adds any toggle switches you'd like, I'll definitely
merge it.

Just don't want to have to spin up a VM and finagle with it.

For what it's worth, I don't see how GetIdbPath or whichever is reporting
an invalid pathname. It should be possible to support your scenario by
asking IDA for the (windows native) path.

**Comment by zachriggle** [comment]
Also, to answer your question from IRC, the `.config` module does expose configuration points.

If you search for other uses, or look at the `config` command in `pwndbg`, you'll see all of the options.

**Comment by disconnect3d** [comment]
Just for an information... on a NTFS filesystem (which I believe is like a default for a machine with Windows) a path to a file that contains ":" is probably treated as NTFS ADS (Alternate Data Stream).

Here is a better screen of this issue:
![ntfs_ads](https://cloud.githubusercontent.com/assets/10009354/20865729/90642560-ba1a-11e6-80f1-8838e9d9ff46.png)

So we have to remove the `:` character from the file path, as it may not work on Windows...
I think I will go with replacing it with `-` or `_`.

More about NTFS ADS:
* [wikipedia NTFS#Alternate_data_streams](https://en.wikipedia.org/wiki/NTFS#Alternate_data_streams_.28ADS.29)
* https://blogs.technet.microsoft.com/askcore/2013/03/24/alternate-data-streams-in-ntfs/

**Comment by disconnect3d** [comment]
Merged in f728a2870a2b22373702cdd81649f1b7b4be9b80 (PR #139)
